E2E test: Multiple consumers reading from the same stream concurrently

## Problem

No E2E test verifies that **multiple consumers** (on separate connections) can read from the same stream independently. Each consumer should receive all messages and maintain independent offsets.

## Expected test

```php
public function testMultipleConsumersOnSameStream(): void
{
    $stream = 'test-multi-consumer-' . uniqid();
    $conn1 = Connection::create(...);
    $conn2 = Connection::create(...);

    $conn1->createStream($stream);

    // Publish messages
    $producer = $conn1->createProducer($stream);
    $producer->sendBatch([...5 AMQP messages...]);
    $producer->waitForConfirms();
    $producer->close();

    // Consumer 1 on conn1
    $consumer1 = $conn1->createConsumer($stream, OffsetSpec::first(), name: 'consumer-1');
    // Consumer 2 on conn2
    $consumer2 = $conn2->createConsumer($stream, OffsetSpec::first(), name: 'consumer-2');

    // Both should receive all 5 messages
    $msgs1 = $this->readAll($consumer1, 5);
    $msgs2 = $this->readAll($consumer2, 5);

    $this->assertCount(5, $msgs1);
    $this->assertCount(5, $msgs2);

    // Store different offsets
    $consumer1->storeOffset($msgs1[2]->getOffset()); // offset at message 3
    $consumer2->storeOffset($msgs2[4]->getOffset()); // offset at message 5

    // Verify independent offsets
    $offset1 = $conn1->queryOffset('consumer-1', $stream);
    $offset2 = $conn2->queryOffset('consumer-2', $stream);
    $this->assertNotSame($offset1, $offset2);
}
```

## Why it matters

- Multiple consumers is the standard pattern for RabbitMQ Streams (fan-out)
- Each consumer must have independent offset tracking
- Tests that consumers on different connections don't interfere with each other

## Acceptance criteria

- [ ] Two consumers on different connections receive all messages independently
- [ ] Each consumer can store and query its own offset
- [ ] Offsets are independent (storing one doesn't affect the other)

## CI Pipeline Failed

Pipeline checks failed. Logs saved to `/home/decodo/work/rabbit-stream/.oda/artifacts/174/pipeline-fail.log`.

```
## Failed CI Checks

Failed: E2E Tests

### Run 23750905054

```
E2E Tests	Run E2E tests	﻿2026-03-30T14:47:48.5584706Z ##[group]Run ./vendor/bin/phpunit --testsuite e2e --testdox
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5585149Z [36;1m./vendor/bin/phpunit --testsuite e2e --testdox[0m
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5611120Z shell: /usr/bin/bash -e {0}
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5611359Z env:
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5611539Z   COMPOSER_PROCESS_TIMEOUT: 0
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5611775Z   COMPOSER_NO_INTERACTION: 1
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5611991Z   COMPOSER_NO_AUDIT: 1
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5612181Z   RABBITMQ_HOST: 127.0.0.1
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5612384Z   RABBITMQ_PORT: 5552
E2E Tests	Run E2E tests	2026-03-30T14:47:48.5612569Z ##[endgroup]
E2E Tests	Run E2E tests	2026-03-30T14:47:48.6421783Z PHPUnit 10.5.63 by Sebastian Bergmann and contributors.
E2E Tests	Run E2E tests	2026-03-30T14:47:48.6422269Z 
E2E Tests	Run E2E tests	2026-03-30T14:47:48.6422404Z Runtime:       PHP 8.3.30
E2E Tests	Run E2E tests	2026-03-30T14:47:48.6422940Z Configuration: /home/runner/work/rabbit-stream/rabbit-stream/phpunit.xml
E2E Tests	Run E2E tests	2026-03-30T14:47:48.6423434Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:20.6804630Z ......F.........................................................S 65 / 82 ( 79%)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4975462Z SS...............                                                 82 / 82 (100%)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4976029Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4976351Z Time: 01:32.852, Memory: 10.00 MB
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4976663Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4977033Z Amqp Message Decoder E2E (CrazyGoat\RabbitStream\Tests\E2E\AmqpMessageDecoderE2E)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.4978302Z  ✔ Decode simple published message
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5000185Z  ✔ Decode message with properties
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5000769Z  ✔ Decode message with application properties
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5001193Z  ✔ Decode multiple messages
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5001340Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5001558Z Close (CrazyGoat\RabbitStream\Tests\E2E\Close)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5002023Z  ✔ Close command
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5002380Z  ✔ Send after client close throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5002720Z  ✘ Send after server close throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5002949Z    │
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5003585Z    │ Failed asserting that 'Failed to read from socket: Success' [Encoding ignored](length: 35) contains "closed" [Encoding ignored](length: 6).
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5004223Z    │
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5004577Z    │ /home/runner/work/rabbit-stream/rabbit-stream/tests/E2E/CloseTest.php:102
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5004947Z    │
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5005142Z  ✔ Double close is idempotent
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5005418Z  ✔ Stream connection send message after close
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5005728Z  ✔ Stream connection read message after close
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5005899Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5006036Z Connection (CrazyGoat\RabbitStream\Tests\E2E\Connection)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5006390Z  ✔ Connection create and stream management
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5006665Z  ✔ Create stream with arguments
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5006891Z  ✔ Get metadata
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5007082Z  ✔ Get stream stats
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5007294Z  ✔ Create duplicate stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5007444Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5008124Z Connection Handshake (CrazyGoat\RabbitStream\Tests\E2E\ConnectionHandshake)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5008620Z  ✔ Peer properties exchange
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5008850Z  ✔ Sasl handshake
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5009060Z  ✔ Sasl authenticate
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5009259Z  ✔ Full handshake
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5009973Z  ✔ Invalid credentials throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5010135Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5010338Z Connection Resilience (CrazyGoat\RabbitStream\Tests\E2E\ConnectionResilience)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5010756Z  ✔ Operation after socket disconnect
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5011058Z  ✔ Is connected returns false after socket error
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5011379Z  ✔ No resource leaks after socket error
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5011542Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5011659Z Consumer (CrazyGoat\RabbitStream\Tests\E2E\Consumer)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5011972Z  ✔ Read returns empty on timeout
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5012219Z  ✔ Produce and consume with read
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5012472Z  ✔ Read one returns single message
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5012718Z  ✔ Store and query offset
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5013043Z  ✔ Auto commit on close stores last offset
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5013350Z  ✔ No auto commit on close does not store offset
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5014033Z  ✔ Auto commit on close with no messages does not store offset
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5014290Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5014454Z Create Stream (CrazyGoat\RabbitStream\Tests\E2E\CreateStream)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5014771Z  ✔ Create stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5015004Z  ✔ Create stream with arguments
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5015257Z  ✔ Create duplicate stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5015438Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5015646Z Create Super Stream (CrazyGoat\RabbitStream\Tests\E2E\CreateSuperStream)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5016017Z  ✔ Create super stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5016274Z  ✔ Create duplicate super stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5016447Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5016620Z Declare Publisher (CrazyGoat\RabbitStream\Tests\E2E\DeclarePublisher)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5017016Z  ✔ Declare publisher with reference
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5017319Z  ✔ Declare publisher without reference
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5017831Z  ✔ Declare publisher on non existent stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5018039Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5018217Z Delete Publisher (CrazyGoat\RabbitStream\Tests\E2E\DeletePublisher)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5018623Z  ✔ Delete publisher after declare
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5018937Z  ✔ Delete non existent publisher throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5019105Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5019256Z Delete Stream (CrazyGoat\RabbitStream\Tests\E2E\DeleteStream)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5019616Z  ✔ Delete stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5019839Z  ✔ Delete non existent stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5019995Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5020168Z Delete Super Stream (CrazyGoat\RabbitStream\Tests\E2E\DeleteSuperStream)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5020576Z  ✔ Delete super stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5020833Z  ✔ Delete non existent super stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5021009Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5021245Z Exchange Command Versions (CrazyGoat\RabbitStream\Tests\E2E\ExchangeCommandVersions)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5021691Z  ✔ Exchange command versions
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5021825Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5021998Z Multiple Consumers (CrazyGoat\RabbitStream\Tests\E2E\MultipleConsumers)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5022413Z  ✔ Multiple consumers on same stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5022566Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5022839Z Named Producer Deduplication (CrazyGoat\RabbitStream\Tests\E2E\NamedProducerDeduplication)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5023345Z  ✔ Named producer deduplication across reconnect
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5023533Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5023758Z Osiris Chunk Parser E2E (CrazyGoat\RabbitStream\Tests\E2E\OsirisChunkParserE2E)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5024180Z  ✔ Parse real chunk from rabbit m q
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5024454Z  ✔ Chunk parser with single message
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5024604Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5024765Z Partitions (CrazyGoat\RabbitStream\Tests\E2E\Partitions)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5025153Z  ✔ Partitions for non existent super stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5025499Z  ✔ Partitions returns streams for super stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5025685Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5025802Z Producer (CrazyGoat\RabbitStream\Tests\E2E\Producer)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5026126Z  ✔ Send and wait for confirms
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5026383Z  ✔ Send batch and wait for confirms
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5026653Z  ✔ Get last publishing id
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5027085Z  ✔ Query sequence for named producer
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5027380Z  ✔ Wait for confirms timeout throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5027789Z  ✔ Producer remains usable after timeout
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5027983Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5028249Z Producer Consumer Offset Resume (CrazyGoat\RabbitStream\Tests\E2E\ProducerConsumerOffsetResume)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5028752Z  ✔ Full lifecycle with offset resume
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5028908Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5029037Z Publish (CrazyGoat\RabbitStream\Tests\E2E\Publish)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5029336Z  ✔ Publish single message
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5029587Z  ✔ Publish multiple messages
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5029719Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5029866Z Publish Error (CrazyGoat\RabbitStream\Tests\E2E\PublishError)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5030311Z  ✔ Publish error callback is invoked when stream deleted
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5030762Z  ✔ Publish error callback is invoked for batch when stream deleted
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5031192Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5031409Z Query Publisher Sequence (CrazyGoat\RabbitStream\Tests\E2E\QueryPublisherSequence)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5031938Z  ✔ Query publisher sequence returns zero for new publisher
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5032318Z  ✔ Query publisher sequence returns last published id
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5032543Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5032647Z Route (CrazyGoat\RabbitStream\Tests\E2E\Route)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5032957Z  ✔ Route for non existent super stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5033277Z  ✔ Route returns streams for routing key
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5033437Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5033687Z Store Offset Query Offset E2E (CrazyGoat\RabbitStream\Tests\E2E\StoreOffsetQueryOffsetE2E)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5034169Z  ✔ Store and query offset at protocol level
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5034476Z  ✔ Query offset for non existent reference
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5034783Z  ✔ Store offset multiple times last value wins
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5034959Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5035132Z Stream Stats (CrazyGoat\RabbitStream\Tests\E2E\StreamStats)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5035486Z  ✔ Stream stats returns statistics
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5035782Z  ✔ Stream stats for non existent stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5035980Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5036108Z Subscribe (CrazyGoat\RabbitStream\Tests\E2E\Subscribe)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5036410Z  ✔ Subscribe to stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5036646Z  ✔ Subscribe with offset last
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5036886Z  ✔ Subscribe with offset next
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5037137Z  ✔ Subscribe to non existent stream throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5037449Z  ✔ Duplicate subscription id throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5037754Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5037894Z Unsubscribe (CrazyGoat\RabbitStream\Tests\E2E\Unsubscribe)
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5038221Z  ✔ Unsubscribe from stream
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5038513Z  ✔ Unsubscribe non existent subscription throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5038846Z  ✔ Unsubscribe already unsubscribed throws
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5039016Z 
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5039107Z FAILURES!
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5039341Z Tests: 82, Assertions: 437, Failures: 1, Skipped: 3.
E2E Tests	Run E2E tests	2026-03-30T14:49:21.5114664Z ##[error]Process completed with exit code 1.
```


```

--- a/tests/E2E/MultipleConsumersTest.php
+++ b/tests/E2E/MultipleConsumersTest.php
@@ -1,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CrazyGoat\RabbitStream\Tests\E2E;
+
+use CrazyGoat\RabbitStream\Client\Connection;
+use CrazyGoat\RabbitStream\Client\Message;
+use CrazyGoat\RabbitStream\Contract\ConsumerInterface;
+use CrazyGoat\RabbitStream\VO\OffsetSpec;
+use PHPUnit\Framework\TestCase;
+
+class MultipleConsumersTest extends TestCase
+{
+    private static string $host = '127.0.0.1';
+    private static int $port = 5552;
+
+    public static function setUpBeforeClass(): void
+    {
+        self::$host = getenv('RABBITMQ_HOST') ?: self::$host;
+        self::$port = (int)(getenv('RABBITMQ_PORT') ?: self::$port);
+    }
+
+    private function amqp(string $body): string
+    {
+        return "\x00\x53\x75\xb0" . pack('N', strlen($body)) . $body;
+    }
+
+    public function testTwoConsumersOnDifferentConnectionsReceiveAllMessages(): void
+    {
+        $stream = 'test-multi-consumer-' . uniqid();
+
+        $conn1 = Connection::create(self::$host, self::$port);
+        $conn2 = Connection::create(self::$host, self::$port);
+
+        try {
+            $conn1->createStream($stream);
+
+            $producer = $conn1->createProducer($stream);
+            $producer->sendBatch([
+                $this->amqp('msg-1'),
+                $this->amqp('msg-2'),
+                $this->amqp('msg-3'),
+                $this->amqp('msg-4'),
+                $this->amqp('msg-5'),
+            ]);
+            $producer->waitForConfirms(timeout: 5);
+            $producer->close();
+
+            $consumer1 = $conn1->createConsumer($stream, OffsetSpec::first(), name: 'consumer-1');
+            $consumer2 = $conn2->createConsumer($stream, OffsetSpec::first(), name: 'consumer-2');
+
+            $msgs1 = $this->readAll($consumer1, 5);
+            $msgs2 = $this->readAll($consumer2, 5);
+
+            $this->assertCount(5, $msgs1);
+            $this->assertCount(5, $msgs2);
+
+            $consumer1->storeOffset($msgs1[2]->getOffset());
+            $consumer2->storeOffset($msgs2[4]->getOffset());
+
+            $offset1 = $conn1->queryOffset('consumer-1', $stream);
+            $offset2 = $conn2->queryOffset('consumer-2', $stream);
+            $this->assertNotSame($offset1, $offset2, 'Each consumer should have an independent offset');
+
+            $consumer1->close();
+            $consumer2->close();
+        } finally {
+            try {
+                $conn1->deleteStream($stream);
+            } catch (\Exception) {
+            }
+            $conn1->close();
+            $conn2->close();
+        }
+    }
+
+    public function testTwoConsumersOnSameConnectionReceiveAllMessages(): void
+    {
+        $stream = 'test-multi-consumer-same-conn-' . uniqid();
+
+        $conn = Connection::create(self::$host, self::$port);
+
+        try {
+            $conn->createStream($stream);
+
+            $producer = $conn->createProducer($stream);
+            $producer->sendBatch([
+                $this->amqp('a'),
+                $this->amqp('b'),
+                $this->amqp('c'),
+            ]);
+            $producer->waitForConfirms(timeout: 5);
+            $producer->close();
+
+            $consumer1 = $conn->createConsumer($stream, OffsetSpec::first(), name: 'shared-conn-c1');
+            $consumer2 = $conn->createConsumer($stream, OffsetSpec::first(), name: 'shared-conn-c2');
+
+            $msgs1 = $this->readAll($consumer1, 3);
+            $msgs2 = $this->readAll($consumer2, 3);
+
+            $this->assertCount(3, $msgs1);
+            $this->assertCount(3, $msgs2);
+
+            $consumer1->close();
+            $consumer2->close();
+        } finally {
+            try {
+                $conn->deleteStream($stream);
+            } catch (\Exception) {
+            }
+            $conn->close();
+        }
+    }
+
+    /** @return Message[] */
+    private function readAll(ConsumerInterface $consumer, int $expectedCount): array
+    {
+        $received = [];
+        $deadline = time() + 10;
+        while (count($received) < $expectedCount && time() < $deadline) {
+            $messages = $consumer->read(timeout: 0.5);
+            foreach ($messages as $msg) {
+                $received[] = $msg;
+            }
+        }
+        return $received;
+    }
+}

